util/fibers/synchronization.h

```diff
@@ -368,104 +369,58 @@ class Done {
+// Because the decrement of Dec() can be observed before notify is called, the counter can be still
+// in use even after Wait() unblocked.
+class EmbeddedBlockingCounter {
+  const uint64_t kCancelFlag = (1ULL << 63);
 
-    bool WaitUntil(const std::chrono::steady_clock::time_point& tp) {
-      uint64_t cnt;
-      std::cv_status status = ec_.await_until(WaitCondition(&cnt), tp);
-      return status == std::cv_status::no_timeout && (cnt & kCancelFlag) == 0;
-    }
-
-    void Cancel() {
-      count_.fetch_or(kCancelFlag, std::memory_order_acq_rel);
-      ec_.notifyAll();
-    }
+  // Re-usable functor for wait condition, stores result in provided pointer
+  auto WaitCondition(uint64_t* cnt) {
+    return [this, cnt]() -> bool {
+      *cnt = count_.load(std::memory_order_relaxed);  // EventCount provides acquire
+      return *cnt == 0 || (*cnt & kCancelFlag);
+    };
+  }
 
-    void Dec() {
-      if (1 == count_.fetch_sub(1, std::memory_order_acq_rel))
-        ec_.notifyAll();
-    }
+ public:
+  EmbeddedBlockingCounter(unsigned start_count = 0) : ec_{}, count_{start_count} {
+  }
 
-   private:
-    friend class BlockingCounter;
-    EventCount ec_;
-    std::atomic<std::uint32_t> use_count_{0};
-    std::atomic<uint64_t> count_;
-  };
+  // Returns true on success (reaching 0), false when cancelled. Acquire semantics
+  bool Wait();
 
-  using ptr_t = ::boost::intrusive_ptr<Impl>;
+  // Same as Wait(), but with timeout
+  bool WaitFor(const std::chrono::steady_clock::duration& duration);
 
- public:
-  explicit BlockingCounter(unsigned count) : impl_(new Impl(count)) {
-  }
+  // Start with specified count. Current value must be zero or cancelled
+  void Start(unsigned cnt);
 
-  void Inc() {
-    Add(1);
-  }
+  // Add to blocking counter
+  void Add(unsigned cnt = 1);
 
-  void Dec() {
-    impl_->Dec();
-  }
+  // Decrement from blocking counter. Release semantics.
+  void Dec();
 
-  void Add(unsigned delta) {
-    impl_->count_.fetch_add(delta, std::memory_order_acq_rel);
-  }
+  // Cancel blocking counter, unblock wait. Release semantics.
+  void Cancel();
 
-  void Cancel() {
-    impl_->Cancel();
-  }
+ private:
+  EventCount ec_;
+  std::atomic<uint64_t> count_;
+};
 
-  // Blocks while the blocking counter is pending.
-  // Returns true on success (reaching 0), false when cancelled.
-  bool Wait() {
-    return impl_->Wait();
-  }
+// A barrier similar to Go's WaitGroup for tracking remote tasks.
+// Internal smart pointer for easier lifetime management. Pass by value.
+struct BlockingCounter {
+  BlockingCounter(unsigned start_count);
 
-  // Blocks while the blocking counter is pending our timeout is reached.
-  // Returns true on success (reaching 0), false when cancelled or timed out.
-  bool WaitFor(const std::chrono::steady_clock::duration& duration) {
-    auto tp = std::chrono::steady_clock::now() + duration;
-    return impl_->WaitUntil(tp);
+  EmbeddedBlockingCounter* operator->() {
```

Comment:
this will require adjusting the whole dragonfly codebase due to interface change.
I suggest you introduce wrappers to preserve the previous interface

Reply:
It's only 3 or 4 usages, I already fixed them. Copying all functions because of this is weird